File: core/broker.py
"""
core/broker.py
Thin wrapper around alpaca-py TradingClient & StockHistoricalDataClient.
All order placement and account queries go through here.
"""

import logging
from datetime import datetime, timedelta, timezone
from typing import Dict, Optional

from alpaca.trading.client import TradingClient
from alpaca.trading.requests import (
    MarketOrderRequest,
    LimitOrderRequest,
    TakeProfitRequest,
    StopLossRequest,
)
from alpaca.trading.enums import OrderSide, TimeInForce, OrderClass
from alpaca.data.historical import StockHistoricalDataClient
from alpaca.data.requests import StockBarsRequest, StockLatestQuoteRequest
from alpaca.data.timeframe import TimeFrame

from config.settings import (
    ALPACA_API_KEY,
    ALPACA_SECRET_KEY,
    PAPER_TRADING,
    STOP_LOSS_PCT,
    TAKE_PROFIT_PCT,
)

log = logging.getLogger(__name__)


class Broker:
    """Handles all Alpaca API interactions."""

    def __init__(self):
        self.trading = TradingClient(
            api_key=ALPACA_API_KEY,
            secret_key=ALPACA_SECRET_KEY,
            paper=PAPER_TRADING,
        )
        self.data = StockHistoricalDataClient(
            api_key=ALPACA_API_KEY,
            secret_key=ALPACA_SECRET_KEY,
        )
        log.info("Broker initialised (paper=%s)", PAPER_TRADING)

    # ── Account ───────────────────────────────────────────────────────────────

    def get_account(self):
        return self.trading.get_account()

    def get_portfolio_value(self) -> float:
        acct = self.get_account()
        return float(acct.portfolio_value)

    def get_buying_power(self) -> float:
        acct = self.get_account()
        return float(acct.buying_power)

    def get_positions(self) -> Dict[str, object]:
        """Returns {symbol: position_object}."""
        positions = self.trading.get_all_positions()
        return {p.symbol: p for p in positions}

    def get_position(self, symbol: str) -> Optional[object]:
        positions = self.get_positions()
        return positions.get(symbol)

    # ── Market Data ───────────────────────────────────────────────────────────

    def get_daily_bars(self, symbol: str, n_days: int = 60):
        """Fetch last n_days of daily bars; returns a pandas DataFrame."""
        start = datetime.now(timezone.utc) - timedelta(days=n_days + 10)
        req = StockBarsRequest(
            symbol_or_symbols=symbol,
            timeframe=TimeFrame.Day,
            start=start,
        )
        bars = self.data.get_stock_bars(req)
        df = bars.df
        if isinstance(df.index, type(None)):
            return df
        # flatten multi-index if necessary
        if hasattr(df.index, "levels"):
            df = df.xs(symbol, level="symbol") if symbol in df.index.get_level_values("symbol") else df
        return df.tail(n_days)

    def get_latest_price(self, symbol: str) -> float:
        req = StockLatestQuoteRequest(symbol_or_symbols=symbol)
        quote = self.data.get_stock_latest_quote(req)
        return float(quote[symbol].ask_price)

    def is_market_open(self) -> bool:
        clock = self.trading.get_clock()
        return clock.is_open

    # ── Orders ────────────────────────────────────────────────────────────────

    def place_bracket_order(
        self,
        symbol: str,
        qty: float,
        side: OrderSide,
        entry_price: float,
    ):
        """
        Submit a market bracket order with automatic stop-loss & take-profit.
        Bracket orders require fractional=False for most leveraged ETFs.
        """
        if side == OrderSide.BUY:
            stop_price   = round(entry_price * (1 - STOP_LOSS_PCT), 2)
            profit_price = round(entry_price * (1 + TAKE_PROFIT_PCT), 2)
        else:
            stop_price   = round(entry_price * (1 + STOP_LOSS_PCT), 2)
            profit_price = round(entry_price * (1 - TAKE_PROFIT_PCT), 2)

        order_data = MarketOrderRequest(
            symbol=symbol,
            qty=int(qty),           # whole shares only for leveraged ETFs
            side=side,
            time_in_force=TimeInForce.DAY,
            order_class=OrderClass.BRACKET,
            take_profit=TakeProfitRequest(limit_price=profit_price),
            stop_loss=StopLossRequest(stop_price=stop_price),
        )

        try:
            order = self.trading.submit_order(order_data)
            log.info(
                "ORDER SUBMITTED | %s %s %d @ ~%.2f | SL=%.2f TP=%.2f | id=%s",
                side.value.upper(), symbol, int(qty), entry_price,
                stop_price, profit_price, order.id,
            )
            return order
        except Exception as exc:
            log.error("Order submission failed for %s: %s", symbol, exc)
            raise

    def close_position(self, symbol: str):
        """Market-close an existing position."""
        try:
            result = self.trading.close_position(symbol)
            log.info("POSITION CLOSED | %s", symbol)
            return result
        except Exception as exc:
            log.error("Failed to close position %s: %s", symbol, exc)
            raise

    def cancel_all_orders(self):
        self.trading.cancel_orders()
        log.info("All open orders cancelled.")

File: app/services/trading/broker.rb
# frozen_string_literal: true

require "erb"
require "json"

module Trading
  class Broker
    class Error < StandardError; end

    def initialize
      raise Error, "Set ALPACA_API_KEY and ALPACA_SECRET_KEY" if Trading.alpaca_api_key.blank? || Trading.alpaca_secret_key.blank?

      @trading_url = Trading.config[:paper_trading] ? "https://paper-api.alpaca.markets" : "https://api.alpaca.markets"
      @data_url = "https://data.alpaca.markets"
    end

    def trading_conn
      @trading_conn ||= Faraday.new(url: @trading_url) do |f|
        f.request :url_encoded
        f.adapter Faraday.default_adapter
        f.headers["APCA-API-KEY-ID"] = Trading.alpaca_api_key
        f.headers["APCA-API-SECRET-KEY"] = Trading.alpaca_secret_key
      end
    end

    def data_conn
      @data_conn ||= Faraday.new(url: @data_url) do |f|
        f.request :url_encoded
        f.adapter Faraday.default_adapter
        f.headers["APCA-API-KEY-ID"] = Trading.alpaca_api_key
        f.headers["APCA-API-SECRET-KEY"] = Trading.alpaca_secret_key
      end
    end

    def get_account
      parse(trading_conn.get("/v2/account"))
    end

    def portfolio_value
      get_account["portfolio_value"].to_f
    end

    def buying_power
      get_account["buying_power"].to_f
    end

    def positions
      rows = parse_array(trading_conn.get("/v2/positions"))
      rows.index_by { |p| p["symbol"] }
    end

    def position(symbol)
      positions[symbol]
    end

    # Returns array of bar hashes: { time:, open:, high:, low:, close:, volume: }
    def daily_bars(symbol, n_days: 60)
      start = (Time.now.utc - (n_days + 15).days).iso8601
      resp = data_conn.get(
        "/v2/stocks/#{ERB::Util.url_encode(symbol)}/bars",
        {
          timeframe: "1Day",
          start: start,
          limit: n_days + 30,
          adjustment: "raw",
          feed: "sip"
        }
      )
      body = parse(resp)
      bars = body["bars"] || []
      normalized = bars.map { |b| normalize_bar(b) }
      normalized.last(n_days)
    end

    def latest_price(symbol)
      resp = data_conn.get("/v2/stocks/#{ERB::Util.url_encode(symbol)}/quotes/latest")
      body = parse(resp)
      quote = body["quote"] || {}
      (quote["ap"] || quote["ask_price"] || quote["bp"] || quote["bid_price"]).to_f
    end

    def market_open?
      body = parse(trading_conn.get("/v2/clock"))
      body["is_open"] == true
    end

    # side: :buy or :sell
    def place_bracket_order(symbol:, qty:, side:, entry_price:)
      cfg = Trading.config
      sl = cfg[:stop_loss_pct].to_f
      tp = cfg[:take_profit_pct].to_f

      stop_price, profit_price =
        if side.to_s == "buy"
          [ (entry_price * (1 - sl)).round(2), (entry_price * (1 + tp)).round(2) ]
        else
          [ (entry_price * (1 + sl)).round(2), (entry_price * (1 - tp)).round(2) ]
        end

      payload = {
        symbol: symbol,
        qty: qty.to_i,
        side: side.to_s,
        type: "market",
        time_in_force: "day",
        order_class: "bracket",
        take_profit: { limit_price: profit_price.to_s },
        stop_loss: { stop_price: stop_price.to_s }
      }

      resp = trading_conn.post("/v2/orders") do |req|
        req.headers["Content-Type"] = "application/json"
        req.body = JSON.generate(payload)
      end
      order = parse(resp)
      Rails.logger.info(
        "ORDER SUBMITTED | #{side.to_s.upcase} #{symbol} #{qty.to_i} @ ~#{entry_price} | SL=#{stop_price} TP=#{profit_price} | id=#{order['id']}"
      )
      order
    end

    def close_position(symbol)
      resp = trading_conn.delete("/v2/positions/#{ERB::Util.url_encode(symbol)}")
      Rails.logger.info("POSITION CLOSED | #{symbol}")
      parse_optional(resp)
    end

    def cancel_all_orders
      resp = trading_conn.delete("/v2/orders")
      Rails.logger.info("All open orders cancelled.")
      parse_optional(resp)
    end

    private

    def parse(resp)
      raise Error, "Alpaca #{resp.status}: #{resp.body}" unless resp.success?

      body = resp.body.to_s
      body.present? ? JSON.parse(body) : {}
    end

    def parse_optional(resp)
      raise Error, "Alpaca #{resp.status}: #{resp.body}" unless resp.success?

      body = resp.body.to_s
      body.present? ? JSON.parse(body) : {}
    end

    def parse_array(resp)
      raise Error, "Alpaca #{resp.status}: #{resp.body}" unless resp.success?

      body = resp.body.to_s
      body.present? ? JSON.parse(body) : []
    end

    def normalize_bar(b)
      {
        time: b["t"],
        open: b["o"].to_f,
        high: b["h"].to_f,
        low: b["l"].to_f,
        close: b["c"].to_f,
        volume: b["v"].to_f
      }
    end
  end
end
